pkg/handler/enqueue.go

```diff
@@ -47,8 +48,11 @@ func (e *EnqueueRequestForObject) Create(evt event.CreateEvent, q workqueue.Rate
 
 // Update implements EventHandler
 func (e *EnqueueRequestForObject) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	// requestSet will help to deduplicate requests in a single event
+	requestSet := reconcile.NewRequestSet()
+
 	if evt.ObjectOld != nil {
-		q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+		requestSet.Insert(reconcile.Request{NamespacedName: types.NamespacedName{
```

Comment:
@mars1024 thanks for your effort! However, the workqueue already deduplicates so this isn't needed: https://github.com/kubernetes/kubernetes/blob/3433f099fb0b5c51d3b549372364f818242f7312/staging/src/k8s.io/client-go/util/workqueue/queue.go#L120-L126

/hold

Reply:
@alvaroaleman Thanks for you reply, but I don't think workqueue could squash all repeated requests into one, as we know, workqueue is not only called by event handler, but also called by controller at the same time.

If workqueue is called as the following steps,
- event handler called `workqueue.Add` with request A, `dirty` set and `queue` slice will have A
- controller called `workqueue.Get` to get A for reconciliation, then `dirty` set and `queue` slice will lose A and `processing` set will have A
- event handler called `workqueue.Add` with request A again in the same update event, `dirty` set will have A again
- controller called `workqueue.Done` after finishing reconciliation A, `processing` set will lose A and `queue` slice will have A because `dirty` set have
- controller called `workqueue.Get` in another loop, then get A for reconciliation again because `queue` slice have
- ...

Just as above, a single update event will trigger two reconciliations for A, IMO, this is unnecessary.

Reply:
friendly ping @alvaroaleman , could you help to take a look at the comments above?

Reply:
I see, yeah it can race with the deqeueing.

Reply:
Can this not be fixed in workqueue?

Reply:
I think the only way to fix that in the workqueue would be to add the ability to take multiple items at once

Reply:
Hi @alvaroaleman , so your final suggestion is fixing this issue in `workqueue` package ? Maybe I can try, but I don't see the difference of implementing deduplication in `workqueue` or `handler`, could you explain this for me?

Reply:
That was a comment for @coderanger and I was referring to the upstream `workqueue` package. I would expect that to cause some debate upstream, though. I don't see how we could get the workqueue itself to deduplicate in such a case safely without atomically taking more than one item.

Downstream, this needs to be in the handler.

Reply:
Understandable that doing it here is probably quicker but this feels a bit off to me, redundant layers that are very similar :) I guess I still don't see the core issue. If the second `Add` happens after the first `Get`, then reconciling twice is correct. We don't know where in the process things would be. In some cases it might be redundant but in the general case?

Reply:
The reconciliation will only look at the state of the object at the time of reconciliation. If we enqueue twice for one event, one of those will end up being a no-op. I agree its a micro optimization. I personally don't have strong feelings one way or another.

Reply:
Personally I prefer to do code changes on `handler`, just because in the old days of `client-go`, `sample-controller` was enqueuing object only once for update event, like [https://github.com/kubernetes/sample-controller/blob/80a6a3c0f86bb65beec9ab6a44fe7ea967ae3ede/controller.go#L131-L140](https://github.com/kubernetes/sample-controller/blob/80a6a3c0f86bb65beec9ab6a44fe7ea967ae3ede/controller.go#L131-L140)
Deduplication seems more reasonable to be done by `handler`.

Reply:
So the goal is just to deal with the short time window between `Get` and the reconcile actually starting?

Reply:
Yeah, if we do add, then reconcile gets it, then add, it gets reconciled twice. If we do both adds before the reconcile gets it, this won't make a difference, as the workqueue will de-duplicate.

Reply:
I mean that overall seems correct. If another event comes in during the middle of a reconcile, we need to reconcile again since that change might not have been fully seen by the running one, depending on the code and how far along it was.

Reply:
@coderanger that is true but the issue here is that one update even might trigger two reconciles if the update event handler enqueues oldObj-> reconciler pops -> update event handler enqueues newObj

A different event must always trigger another reconciliation, that is correct.

Reply:
I've never really understood why we enqueue ObjOld since by definition it shouldn't be relevant. The only case where it would matter is detecting object renames which Kubernetes doesn't support anyway. Maybe some kind of edge case with forced overwrites and finalizers?

Reply:
But in a more concrete answer, an `AddMany` API in workqueue seems like a reasonable thing to add :) Would be the same as Add but a `...` args list (or even just make Add take ...) and make it atomic-y? Might be literally the same code as this but inside Add.

Reply:
> I've never really understood why we enqueue ObjOld since by definition it shouldn't be relevant. The only case where it would matter is detecting object renames which Kubernetes doesn't support anyway. Maybe some kind of edge case with forced overwrites and finalizers?

Yeah I wasn't entirely sure about that, does the kube api guarantee that updates have an old object and a new object?

> But in a more concrete answer, an AddMany API in workqueue seems like a reasonable thing to add :) Would be the same as Add but a ... args list (or even just make Add take ...) and make it atomic-y? Might be literally the same code as this but inside Add.

Give it a stab upstream? :)

Reply:
> Yeah I wasn't entirely sure about that, does the kube api guarantee that updates have an old object and a new object?

Even if they didn't, simpler code would be more like "if new enqueue new else if old enqueue old else wat". That would be a lot less tracking stuff in here, and removes allocating a set on each event.

Reply:
Yeah the more I look at this code, there is no reason why any event should ever result in more than one reconcile request. The existing code is just wrong, or is trying to plan for impossible events (like object renames or something).

Reply:
Also both the Owner and Mapped handlers do similar thing, though there it's more defensible since you can change an owner in an update and reconciling both seems reasonable. But that would need yet more complexity to do the fast check >_<

Reply:
Emmm... So a clear and precise solution hasn't came up, `if (new) {enqueue new} else if (old) {enqueue old} else {wat}` seems a good way for Object handler, but for Owner and Mapped handlers, both old and new objects should be handled, ref to [pod update event handler in replica set controller](https://github.com/kubernetes/kubernetes/blob/c417626c4912161eb9aecafe53e8b081b7105e7b/pkg/controller/replicaset/replica_set.go#L399-L469). We have to admit that a single update event can trigger one or several requests.

Reply:
This PR doesn't change owner and mapping handler, so why are we debating them? I like the idea of @coderanger to just check upstream if they are ok changing the workqueue to allow adding multiple items at once, that is by far the most elegant solution of all

Reply:
> This PR doesn't change owner and mapping handler, so why are we debating them?

Because those have exactly the same issue and if we fix it for one, we should probably fix it for all of them? If we were going to move forward with this solution, it would need to be applied there too :)

Reply:
> > This PR doesn't change owner and mapping handler, so why are we debating them?
> 
> Because those have exactly the same issue and if we fix it for one, we should probably fix it for all of them? If we were going to move forward with this solution, it would need to be applied there too :)

Yes, I didn't do some code changes on other handlers is not because they don't have similar issues, is just because i'm not sure that this is really an issue (of course, we believe that it is now) ...

@alvaroaleman @coderanger Could you help to invoke more members for a discussion or a decision on changing`workqueue` package?

Reply:
I'm working on a PR for client-go now. Or rather working out how to test it since that's the hard part :)

Reply:
https://github.com/kubernetes/kubernetes/pull/99185 filed.

Reply:
```
// Update implements EventHandler
func (e *EnqueueRequestForObject) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
	var oldReq reconcile.Request
	if evt.MetaOld != nil {
		oldReq = reconcile.Request{NamespacedName: types.NamespacedName{
			Name:      evt.MetaOld.GetName(),
			Namespace: evt.MetaOld.GetNamespace(),
		}}
		q.Add(oldReq)
	} else {
		enqueueLog.Error(nil, "UpdateEvent received with no old metadata", "event", evt)
	}

	if evt.MetaNew != nil {
		newReq := reconcile.Request{NamespacedName: types.NamespacedName{
			Name:      evt.MetaNew.GetName(),
			Namespace: evt.MetaNew.GetNamespace(),
		}}
		if oldReq != newReq {
			q.Add(newReq)
		}
	} else {
		enqueueLog.Error(nil, "UpdateEvent received with no new metadata", "event", evt)
	}
}
```

Reply:
@sttts the upstream change isn't really wanted because of this handler, as you noted, fixing it here is trivial and won't change anything about the number of allocations we do.

It is about something like this https://github.com/kubernetes-sigs/controller-runtime/blob/2faebdb710c1626efaa1e4df7cfb05c32ce1bdf5/pkg/handler/enqueue_mapped.go#L61-L62

There, we now suddenly need to allocate a new set for every update event to build deduplication in front of the already existing set(s) that is the workqueue. Having to build an ephemeral second set because the first set doesn't allow to atomically insert multiple entries seems needlessly inefficient.